File: apps/web/src/agent/prompt.ts
import { Catalog, CatalogEntry } from "@mind-imprint/contracts";
import { ChatTool } from "../llm/types";

export const DEMO_TWINS = ["sift", "craap", "steelman"] as const;

export function buildCatalogText(catalog: Catalog): string {
  // Group entries by category
  const groups = new Map<string, CatalogEntry[]>();
  for (const entry of catalog) {
    const existing = groups.get(entry.category);
    if (existing) {
      existing.push(entry);
    } else {
      groups.set(entry.category, [entry]);
    }
  }

  const lines: string[] = [];
  for (const [category, entries] of groups) {
    lines.push(`【${category}】`);
    for (const e of entries) {
      const kind = e.interaction_type ? `（${e.interaction_type}）` : "";
      lines.push(`· ${e.id}｜${e.name}${kind}`);
      lines.push(`   何时用：${e.trigger_condition}`);
      lines.push(`   能帮他：${e.purpose}`);
    }
  }
  return lines.join("\n");
}

export function buildSystemPrompt(catalog: Catalog): string {
  const catalogText = buildCatalogText(catalog);
  const PROMPT_TEMPLATE = `# 角色
你是「思维印记」里的思维陪练——更像一位**导师 / 教练**，服务国际课程（IB）方向的学生。学生带着自己真实的任务（论文、项目、课题、阅读）来。你的价值不是当一台答案机，而是在协作中把「思考」交回给他自己，让他离开时比来时更会想。

# 你怎么帮（克制，但不是只会反问）
- **不替他定论、不替他写、不替他判对错好坏。** 该他想的，别替他想完。
- 你有一整套教练手段，按情况挑用，而不是每次都反问：
  - 给一个**提示**，把他往前推一小步；
  - 问一个**引导性问题**，让他自己发现缺口；
  - **指出一个他没注意到的角度**或可能的反例；
  - **肯定**他已经做对的部分，让他知道哪条路走对了；
  - 必要时，**提议一张思维工具卡**（见下，按需，不是默认动作）。
- **聚焦一步。** 一次只推进一个焦点，简短、口语；别一口气抛一堆问题或长篇大论——保护他的思考节奏。
- **善用排版。** 用 Markdown 让重点一眼可见：\`**加粗**\`关键词，必要时配小标题 / 列表 / \`>\` 引用。突出重点，但整体仍简短。

# 关于链接和外部资料（重要）
你**打不开链接、也看不到网页或文件里的内容**——你只看得到学生在对话里贴出的文字。所以当学生只丢来一个链接（或提到某个网页/PDF）时：
- **别假装读过它**，别凭标题或网址猜测、编造里面的内容。
- 坦诚说明你看不到链接内容，请他把**关键段落 / 数据 / 原话**粘贴进来；或者用一两句话先讲讲他从中看到了什么。
- 这正好是个起点：可以借机和他一起**溯源、核实**这份材料（必要时再提议相应的工具卡）。

# 工具卡（贴合就递，别犹豫）
工具卡是你手里一种有力的手段。当此刻的处境**贴合**某张卡时，把它递给他就是好陪练——别因为「怕越界」就压着不给。
- 目录里每张卡都标了「何时用」(适用情形) 和「能帮他」(这张卡能给学生什么)。当学生此刻的处境贴合某卡这两栏时，就用 \`summon_card\` 提议它——这不违反克制；克制是指不替他定论，**不是把工具藏起来**。
- **一次最多一张**；同时贴合多张时，挑最综合 / 最贴合当前任务的那张。真的没有贴合的卡，就正常陪练，别硬塞。
- 先按**分类**判断他现在卡在哪一类问题上，再在该类里挑最贴合的那一张。
- \`reason\` 写给系统看（为什么此刻贴合）；\`nudge_text\` 写给学生看（一句自然、邀请式、不命令的话）。**打开由学生确认**——你只是提议。
- 学生**婉拒 / 跳过**一张卡时，尊重他，继续陪练，**不要反复弹**同一张卡。
- 学生**提交**一张卡后，你会拿到他填写内容的结构化结果。基于他**自己写下的**东西继续——先接住他的思考，再就其中**一处**往前推一步。

# 可用的思维工具卡目录（按分类）
{{catalog}}`;

  return PROMPT_TEMPLATE.replace("{{catalog}}", catalogText);
}

export function demoCatalog(full: Catalog): Catalog {
  return full.filter((c) => !(DEMO_TWINS as readonly string[]).includes(c.id));
}

export function summonCardTool(catalog: Catalog): ChatTool {
  return {
    name: "summon_card",
    description: "当此刻贴合某卡的适用情形时，提议这张卡。一次最多一张。",
    parameters: {
      type: "object",
      properties: {
        card_id: {
          type: "string",
          enum: catalog.map((c) => c.id),
        },
        reason: {
          type: "string",
        },
        nudge_text: {
          type: "string",
        },
      },
      required: ["card_id", "reason", "nudge_text"],
    },
  };
}

File: apps/api/internal/agent/prompt.go
// Package agent is the server-side agent brain ported from the TS app:
// the 克制阶梯 system prompt, the summon_card tool, refeed serialization, the
// LLM message-history mapping, and the turn-loop engine. Golden fixtures
// in testdata/ track intentional changes to the assembled prompt.
package agent

import (
	"strconv"
	"strings"

	"mindimprint/api/internal/cards"
	"mindimprint/api/internal/gateway"
)

// promptTemplate is the legacy coach instruction. BuildSystemPrompt replaces
// {{catalog}} with the registry-derived card directory.
const promptTemplate = `# 角色
你是“思维印记”的学习陪练，帮助国际课程学生理解材料、检验思路并完成自己的学习任务。回复直接展示给学生，用“你”称呼学生，清楚说明当前讨论的内容。

# 提供帮助
- 学生自己作出判断、撰写要提交的正文。你可以解释概念、指出已有论述中的依据或缺口、提供提示和引导问题，不代写正文或替学生作出最终决定。
- 根据学生当前需要选择帮助方式：解释一个概念、提出一个具体问题、提示相关条件或反例，或说明已经完成的某项思考。肯定与建议都应有具体依据。
- 一轮聚焦一个问题，使用自然、完整的句子；需要多项说明时用简短列表。无需在每次解释后追加问题。
- 可用 Markdown 标出重点，避免过多标题和强调。

# 材料范围
你只能依据对话和系统提供的材料，不能自行打开链接或文件。学生只给出链接、标题或文件名时，不假装读过正文，也不推测内容；请学生粘贴与问题相关的段落、数据或原话。讨论来源时区分已提供的信息和仍需核实的内容。

# 工具卡
- 根据目录中的适用情形和用途选择工具卡；当前任务与某张卡相符时，调用 summon_card 提议使用。
- 一次最多提议一张；多张均适用时选择最相关的一张，没有适用卡片时继续对话。
- reason 说明本次选择的依据，供系统使用；nudge_text 直接给学生看，用一句话说明卡片能帮助处理什么问题，并邀请学生使用。
- 工具卡由学生确认打开。学生跳过或拒绝后继续提供帮助，不重复弹出同一张卡。
- 学生提交工具卡后，依据实际填写内容回应，并围绕其中一个需要继续思考的地方提供帮助。

# 可用的思维工具卡目录（按分类）
{{catalog}}`

// Material is the agent's view of a task material (see slice-2 material table).
type Material struct {
	ID     string
	Title  string
	Blocks []MaterialBlock
}

// MaterialBlock is one addressable paragraph of a material.
type MaterialBlock struct {
	ID   string `json:"id"`
	Text string `json:"text"`
}

// materialAlias returns the prompt-local alias for the i-th material in a
// materials slice: "m0", "m1", … — the index, NOT m.ID. Block ids are only
// "stable within a material" (materialize.Segment), so a project with ≥2
// materials needs SOME qualifier to disambiguate colliding bare [b0] labels.
// The qualifier can't be m.ID itself: on the course path m.ID is the literal
// "m0", but on the studio path (api/studioturn.go's projectMaterials) it is
// the material's full 36-char UUID — and the L1 instruction's own example
// ("m0:b0", buildAnchorPrompt) is only unmistypeable by a beginner student
// AND by the model when the label is this short. blockLookup (anchors.go)
// derives the identical alias from the SAME materials slice in the SAME
// order, so the pair stays in lockstep without a shared table or a migration.
func materialAlias(i int) string {
	return "m" + strconv.Itoa(i)
}

// BuildMaterialContext renders the task's materials (with block ids) as a
// context block the model can reference when anchoring questions. Empty when
// there are no materials.
func BuildMaterialContext(materials []Material) string {
	if len(materials) == 0 {
		return ""
	}
	var b strings.Builder
	b.WriteString("# 学生正在读/写的材料（可引用其中的 block_id 与原句）\n")
	for i, m := range materials {
		alias := materialAlias(i)
		b.WriteString("## 材料：" + m.Title + "\n")
		for _, blk := range m.Blocks {
			b.WriteString("[" + alias + ":" + blk.ID + "] " + blk.Text + "\n")
		}
	}
	return b.String()
}

// BuildCatalogText groups catalog entries by category (first-seen insertion
// order) and renders the Chinese lines exactly as the TS buildCatalogText does.
func BuildCatalogText(catalog []cards.Spec) string {
	var order []string
	byCat := map[string][]cards.Spec{}
	for _, e := range catalog {
		if _, seen := byCat[e.Category]; !seen {
			order = append(order, e.Category)
		}
		byCat[e.Category] = append(byCat[e.Category], e)
	}

	var lines []string
	for _, category := range order {
		lines = append(lines, "【"+category+"】")
		for _, e := range byCat[category] {
			kind := ""
			if e.InteractionType != "" {
				kind = "（" + e.InteractionType + "）"
			}
			lines = append(lines, "· "+e.ID+"｜"+e.Name+kind)
			lines = append(lines, "   何时用："+e.TriggerCondition)
			lines = append(lines, "   用途："+e.Purpose)
		}
	}
	return strings.Join(lines, "\n")
}

// BuildSystemPrompt fills the {{catalog}} placeholder in the 克制阶梯 template.
func BuildSystemPrompt(catalog []cards.Spec) string {
	return strings.Replace(promptTemplate, "{{catalog}}", BuildCatalogText(catalog), 1)
}

// SummonCardTool builds the summon_card ChatTool whose card_id enum is the
// catalog ids. Mirrors the TS summonCardTool exactly (description + required).
func SummonCardTool(catalog []cards.Spec) gateway.ChatTool {
	ids := make([]string, len(catalog))
	for i, c := range catalog {
		ids[i] = c.ID
	}
	return gateway.ChatTool{
		Name:        "summon_card",
		Description: "当此刻贴合某卡的适用情形时，提议这张卡。一次最多一张。",
		Parameters: map[string]any{
			"type": "object",
			"properties": map[string]any{
				"card_id": map[string]any{
					"type": "string",
					"enum": ids,
				},
				"reason": map[string]any{
					"type": "string",
				},
				"nudge_text": map[string]any{
					"type": "string",
				},
			},
			"required": []string{"card_id", "reason", "nudge_text"},
		},
	}
}
